h-include.js

```diff
@@ -41,35 +41,64 @@ var hinclude;
   hinclude = {
     classprefix: "include_",
 
-    show_content: function (element, req) {
-      var i, include, message, fragment = element.getAttribute('fragment') || 'body';
-      if (req.status === 200 || req.status === 304) {
-        var container = document.implementation.createHTMLDocument().documentElement;
-        container.innerHTML = req.responseText;
+    create_container: function (req) {
+      var container = document.implementation.createHTMLDocument().documentElement;
+      container.innerHTML = req.responseText;
+
+      return container;
+    },
+    check_recursion: function (container, element) {
+      // Check for recursion against current browser location
+      if(element.getAttribute('src') === document.location.href) {
+        throw new Error('Recursion not allowed');
+      }
 
-        var includes = container.getElementsByTagName('h-include');
-        for (i = 0; i < includes.length; i = i + 1) {
-          include = includes[i];
+      // Check for recursion is ascendents
+      var elementToCheck = element.parentNode;
+      while (elementToCheck.parentNode) {
+        if (elementToCheck.nodeName === 'H-INCLUDE') {
 
-          if (include.getAttribute('src') === element.getAttribute('src')) {
-            message = "Saw nested h-include with same src as ancestor (recursion).";
-            element.innerHTML = message;
-            throw new Error(message);
+          if (element.getAttribute('src') === elementToCheck.getAttribute('src')) {
+            throw new Error('Recursion not allowed');
           }
         }
 
-        var node = container.querySelector(fragment);
+        elementToCheck = elementToCheck.parentNode;
+      }
+    },
+    extract_fragment: function (container, fragment, req) {
+      var node = container.querySelector(fragment);
 
-        if (!node) {
-          console.warn("Did not find fragment in response");
-          return;
-        }
+      if (!node) {
+        throw new Error("Did not find fragment in response");
+      }
 
-        element.innerHTML = node.innerHTML;
-        
-        element.onSuccess && element.onSuccess();
+      return node;
+    },
+    replace_content: function (element, node) {
+      element.innerHTML = node.innerHTML;
+    },
+    on_end: function (element, req) {
+      element.className = 'included ' + hinclude.classprefix + req.status;
+    },
+
+    show_content: function (element, req) {
+      var fragment = element.getAttribute('fragment') || 'body';
+      if (req.status === 200 || req.status === 304) {
+        var createContainer = element.createContainer || hinclude.create_container;
+        var container = createContainer(req);
+
+        hinclude.check_recursion(container, element);
```

Comment:
Shouldn't `check_recursion()` be called in `include()` rather than `show_content()`?